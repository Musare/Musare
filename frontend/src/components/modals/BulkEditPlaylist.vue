<script setup lang="ts">
import { reactive, computed, defineAsyncComponent } from "vue";
import Toast from "toasters";
import { useWebsocketsStore } from "@/stores/websockets";
import { useLongJobsStore } from "@/stores/longJobs";
import { useModalsStore } from "@/stores/modals";

const Modal = defineAsyncComponent(() => import("@/components/Modal.vue"));
const PlaylistItem = defineAsyncComponent(
	() => import("@/components/PlaylistItem.vue")
);
const QuickConfirm = defineAsyncComponent(
	() => import("@/components/QuickConfirm.vue")
);

const props = defineProps({
	modalUuid: { type: String, required: true },
	mediaSources: { type: Array, required: true }
});

const { closeCurrentModal } = useModalsStore();

const { setJob } = useLongJobsStore();

const { socket } = useWebsocketsStore();

const { openModal } = useModalsStore();

const search = reactive({
	query: "",
	searchedQuery: "",
	page: 0,
	count: 0,
	resultsLeft: 0,
	pageSize: 0,
	results: []
});

const resultsLeftCount = computed(() => search.count - search.results.length);

const nextPageResultsCount = computed(() =>
	Math.min(search.pageSize, resultsLeftCount.value)
);

const searchForPlaylists = page => {
	if (search.page >= page || search.searchedQuery !== search.query) {
		search.results = [];
		search.page = 0;
		search.count = 0;
		search.resultsLeft = 0;
		search.pageSize = 0;
	}

	const { query } = search;
	const action = "playlists.searchAdmin";

	search.searchedQuery = search.query;
	socket.dispatch(action, query, page, res => {
		const { data } = res;
		if (res.status === "success") {
			const { count, pageSize, playlists } = data;
			search.results = [...search.results, ...playlists];
			search.page = page;
			search.count = count;
			search.resultsLeft = count - search.results.length;
			search.pageSize = pageSize;
		} else if (res.status === "error") {
			search.results = [];
			search.page = 0;
			search.count = 0;
			search.resultsLeft = 0;
			search.pageSize = 0;
			new Toast(res.message);
		}
	});
};

const addSongsToPlaylist = playlistId => {
	let id;
	let title;

	socket.dispatch(
		"playlists.addSongsToPlaylist",
		playlistId,
		props.mediaSources,
		{
			cb: () => {},
			onProgress: res => {
				if (res.status === "started") {
					id = res.id;
					title = res.title;
					closeCurrentModal();
				}

				if (id)
					setJob({
						id,
						name: title,
						...res
					});
			}
		}
	);
};

const removeSongsFromPlaylist = playlistId => {
	let id;
	let title;

	socket.dispatch(
		"playlists.removeSongsFromPlaylist",
		playlistId,
		props.mediaSources,
		{
			cb: data => {
				console.log("FINISHED", data);
			},
			onProgress: res => {
				if (res.status === "started") {
					id = res.id;
					title = res.title;
					closeCurrentModal();
				}

				if (id)
					setJob({
						id,
						name: title,
						...res
					});
			}
		}
	);
};
</script>

<template>
	<div>
		<modal
			title="Bulk Edit Playlist"
			class="bulk-edit-playlist-modal"
			size="slim"
		>
			<template #body>
				<div>
					<label class="label">Search for a playlist</label>
					<div class="control is-grouped input-with-button">
						<p class="control is-expanded">
							<input
								class="input"
								type="text"
								placeholder="Enter your playlist query here..."
								v-model="search.query"
								@keyup.enter="searchForPlaylists(1)"
							/>
						</p>
						<p class="control">
							<a
								class="button is-info"
								@click="searchForPlaylists(1)"
								><i class="material-icons icon-with-button"
									>search</i
								>Search</a
							>
						</p>
					</div>
					<div v-if="search.results.length > 0">
						<playlist-item
							v-for="playlist in search.results"
							:key="`searchKey-${playlist._id}`"
							:playlist="playlist"
							:show-owner="true"
						>
							<template #actions>
								<quick-confirm
									@confirm="addSongsToPlaylist(playlist._id)"
								>
									<i
										class="material-icons add-to-playlist-icon"
										:content="`Add songs to playlist`"
										v-tippy
									>
										playlist_add
									</i>
								</quick-confirm>
								<quick-confirm
									@confirm="
										removeSongsFromPlaylist(playlist._id)
									"
								>
									<i
										class="material-icons remove-from-playlist-icon"
										:content="`Remove songs from playlist`"
										v-tippy
									>
										playlist_remove
									</i>
								</quick-confirm>
								<i
									@click="
										openModal({
											modal: 'editPlaylist',
											props: { playlistId: playlist._id }
										})
									"
									class="material-icons edit-icon"
									content="Edit Playlist"
									v-tippy
									>edit</i
								>
							</template>
						</playlist-item>
						<button
							v-if="resultsLeftCount > 0"
							class="button is-primary load-more-button"
							@click="searchForPlaylists(search.page + 1)"
						>
							Load {{ nextPageResultsCount }} more results
						</button>
					</div>
				</div>
			</template>
		</modal>
	</div>
</template>

<style lang="less" scoped>
.label {
	text-transform: capitalize;
}

.playlist-item:not(:last-of-type) {
	margin-bottom: 10px;
}
.load-more-button {
	width: 100%;
	margin-top: 10px;
}
</style>
