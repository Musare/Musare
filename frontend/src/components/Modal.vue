<script setup lang="ts">
import { defineAsyncComponent, ref, onMounted } from "vue";
import { useModalsStore } from "@/stores/modals";

const ChristmasLights = defineAsyncComponent(
	() => import("@/components/ChristmasLights.vue")
);

defineProps({
	title: { type: String, default: "Modal" },
	size: { type: String, default: null },
	split: { type: Boolean, default: false }
});

const christmas = ref(false);

const { closeCurrentModal } = useModalsStore();

onMounted(async () => {
	christmas.value = await lofig.get("siteSettings.christmas");
});
</script>

<template>
	<div class="modal is-active">
		<div class="modal-background" @click="closeCurrentModal()" />
		<slot name="sidebar" />
		<div
			:class="{
				'modal-card': true,
				'modal-slim': size === 'slim',
				'modal-wide': size === 'wide',
				'modal-split': split
			}"
		>
			<header class="modal-card-head">
				<slot name="toggleMobileSidebar" />
				<h2 class="modal-card-title is-marginless">
					{{ title }}
				</h2>
				<span class="delete material-icons" @click="closeCurrentModal()"
					>highlight_off</span
				>
				<Transition>
					<christmas-lights v-if="christmas" small :lights="5" />
				</Transition>
			</header>
			<section class="modal-card-body">
				<slot name="body" />
			</section>
			<footer
				:class="{
					'modal-card-foot': true,
					blank: $slots['footer'] == null
				}"
			>
				<slot name="footer" />
			</footer>
		</div>
	</div>
</template>

<style lang="less">
.night-mode .modal .modal-card {
	.modal-card-head,
	.modal-card-foot {
		background-color: var(--dark-grey-3);
		border: none;
	}

	.modal-card-body {
		background-color: var(--dark-grey-4) !important;
	}

	.modal-card-head .delete.material-icons,
	.modal-card-title {
		color: var(--white);
	}

	p,
	label,
	td,
	th {
		color: var(--light-grey-2) !important;
	}

	h1,
	h2,
	h3,
	h4,
	h5,
	h6 {
		color: var(--white) !important;
	}
}

.modal {
	display: flex;
	position: fixed;
	top: 0;
	bottom: 0;
	left: 0;
	right: 0;
	z-index: 1984;
	justify-content: center;
	align-items: center;

	.modal-background {
		position: absolute;
		top: 0;
		bottom: 0;
		left: 0;
		right: 0;
		background-color: rgba(10, 10, 10, 0.85);
	}

	.modal-card {
		display: flex;
		flex-direction: column;
		position: relative;
		width: 800px;
		max-width: calc(100% - 40px);
		max-height: calc(100vh - 40px);
		overflow: visible;
		margin: 0;
		font-size: 16px;

		&.modal-slim {
			width: 640px;
		}

		&.modal-wide {
			width: 1300px;
		}

		&.modal-split {
			height: 100%;

			.modal-card-body {
				display: flex;
				flex-wrap: wrap;
				height: 100%;
				row-gap: 24px;

				.left-section,
				.right-section {
					flex-basis: 50%;
					max-height: 100%;
					overflow-y: auto;
					flex-grow: 1;

					.section {
						display: flex;
						flex-direction: column;
						flex-grow: 1;
						width: auto;
						padding: 15px !important;
						margin: 0 10px;
					}

					@media screen and (max-width: 1100px) {
						flex-basis: 100%;
						max-height: unset;
					}
				}
			}
		}

		.modal-card-head,
		.modal-card-foot {
			display: flex;
			flex-shrink: 0;
			position: relative;
			justify-content: flex-start;
			align-items: center;
			padding: 20px;
			background-color: var(--light-grey);
		}

		.modal-card-head {
			border-bottom: 1px solid var(--light-grey-2);
			border-radius: @border-radius @border-radius 0 0;

			.modal-card-title {
				display: flex;
				flex: 1;
				margin: 0;
				font-size: 26px;
				font-weight: 600;
			}

			.delete.material-icons {
				font-size: 28px;
				cursor: pointer;
				user-select: none;
				-webkit-user-drag: none;
				&:hover,
				&:focus {
					filter: brightness(90%);
				}
			}
		}

		.modal-card-foot {
			border-top: 1px solid var(--light-grey-2);
			border-radius: 0 0 @border-radius @border-radius;
			overflow-x: auto;
			column-gap: 16px;

			& > div {
				display: flex;
				flex-grow: 1;
				column-gap: 16px;
			}

			.right {
				display: flex;
				margin-left: auto;
				margin-right: 0;
				justify-content: flex-end;
				column-gap: 16px;
			}

			&.blank {
				padding: 10px;
			}
		}

		.modal-card-body {
			flex: 1;
			flex-wrap: wrap;
			padding: 20px;
			overflow: auto;
			background-color: var(--white);
		}

		@media screen and (max-width: 650px) {
			max-height: 100vh;
			height: 100%;
			max-width: 100%;
			.modal-card-head,
			.modal-card-foot {
				border-radius: 0;
			}
		}
	}
}

.christmas-mode .modal .modal-card-head .christmas-lights {
	top: 69px !important;
}
</style>
