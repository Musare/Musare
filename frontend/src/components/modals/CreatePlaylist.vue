<script setup lang="ts">
import { defineAsyncComponent, ref, onBeforeUnmount } from "vue";
import Toast from "toasters";
import validation from "@/validation";
import { useWebsocketsStore } from "@/stores/websockets";
import { useModalsStore } from "@/stores/modals";

const Modal = defineAsyncComponent(() => import("@/components/Modal.vue"));

const props = defineProps({
	modalUuid: { type: String, required: true },
	admin: { type: Boolean, default: false }
});

const playlist = ref({
	displayName: "",
	privacy: "public",
	songs: []
});

const { openModal, closeCurrentModal } = useModalsStore();

const { socket } = useWebsocketsStore();

const createPlaylist = () => {
	const { displayName } = playlist.value;

	if (!validation.isLength(displayName, 1, 64))
		return new Toast("Display name must have between 1 and 64 characters.");
	if (!validation.regex.ascii.test(displayName))
		return new Toast(
			"Invalid display name format. Only ASCII characters are allowed."
		);

	return socket.dispatch(
		"playlists.create",
		{ ...playlist.value, admin: props.admin },
		res => {
			new Toast(res.message);

			if (res.status === "success") {
				closeCurrentModal();

				if (!window.addToPlaylistDropdown) {
					openModal({
						modal: "editPlaylist",
						props: { playlistId: res.data.playlistId }
					});
				}
			}
		}
	);
};

onBeforeUnmount(() => {
	if (window.addToPlaylistDropdown)
		window.addToPlaylistDropdown.tippy.setProps({
			zIndex: 9999,
			hideOnClick: true
		});

	window.addToPlaylistDropdown = null;
});
</script>

<template>
	<modal
		:title="admin ? 'Create Admin Playlist' : 'Create Playlist'"
		:size="'slim'"
	>
		<template #body>
			<p class="control is-expanded">
				<label class="label">Display Name</label>
				<input
					v-model="playlist.displayName"
					class="input"
					type="text"
					placeholder="Enter display name..."
					autofocus
					@keyup.enter="createPlaylist()"
				/>
			</p>

			<div class="control" id="privacy-selection">
				<label class="label">Privacy</label>
				<p class="control is-expanded select">
					<select v-model="playlist.privacy">
						<option value="private">Private</option>
						<option value="public" selected>Public</option>
					</select>
				</p>
			</div>
		</template>
		<template #footer>
			<a class="button is-info" @click="createPlaylist()"
				><i class="material-icons icon-with-button">create</i>Create
				Playlist</a
			>
		</template>
	</modal>
</template>
