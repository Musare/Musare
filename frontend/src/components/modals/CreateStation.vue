<script setup lang="ts">
import { defineAsyncComponent, ref } from "vue";
import Toast from "toasters";
import { storeToRefs } from "pinia";
import { useWebsocketsStore } from "@/stores/websockets";
import { useModalsStore } from "@/stores/modals";
import { useUserPreferencesStore } from "@/stores/userPreferences";
import validation from "@/validation";

const Modal = defineAsyncComponent(() => import("@/components/Modal.vue"));

const props = defineProps({
	modalUuid: { type: String, required: true },
	official: { type: Boolean, default: false }
});

const { socket } = useWebsocketsStore();

const { closeCurrentModal } = useModalsStore();

const userPreferencesStore = useUserPreferencesStore();
const { defaultStationPrivacy } = storeToRefs(userPreferencesStore);

const newStation = ref({
	name: "",
	displayName: "",
	description: "",
	privacy: defaultStationPrivacy.value
});

const submitModal = () => {
	newStation.value.name = newStation.value.name.toLowerCase();
	const { name, displayName, description, privacy } = newStation.value;

	if (!name || !displayName || !description)
		return new Toast("Please fill in all fields");

	if (!validation.isLength(name, 2, 16))
		return new Toast("Name must have between 2 and 16 characters.");

	if (!validation.regex.az09_.test(name))
		return new Toast(
			"Invalid name format. Allowed characters: a-z, 0-9 and _."
		);

	if (!validation.isLength(displayName, 2, 32))
		return new Toast("Display name must have between 2 and 32 characters.");
	if (!validation.regex.ascii.test(displayName))
		return new Toast(
			"Invalid display name format. Only ASCII characters are allowed."
		);

	if (!validation.isLength(description, 2, 200))
		return new Toast("Description must have between 2 and 200 characters.");

	let characters = description.split("");

	characters = characters.filter(
		character => character.charCodeAt(0) === 21328
	);

	if (characters.length !== 0)
		return new Toast("Invalid description format.");

	return socket.dispatch(
		"stations.create",
		{
			name,
			type: props.official ? "official" : "community",
			displayName,
			description,
			privacy
		},
		res => {
			if (res.status === "success") {
				new Toast(`You have added the station successfully`);
				closeCurrentModal();
			} else new Toast(res.message);
		}
	);
};
</script>

<template>
	<modal
		:title="
			official ? 'Create Official Station' : 'Create Community Station'
		"
	>
		<template #body>
			<label class="label">Name (unique lowercase station id)</label>
			<p class="control">
				<input
					v-model="newStation.name"
					class="input station-id"
					type="text"
					placeholder="Name..."
					autofocus
				/>
			</p>
			<label class="label">Display Name</label>
			<p class="control">
				<input
					v-model="newStation.displayName"
					class="input"
					type="text"
					placeholder="Display name..."
				/>
			</p>
			<label class="label">Description</label>
			<p class="control">
				<input
					v-model="newStation.description"
					class="input"
					type="text"
					placeholder="Description..."
				/>
			</p>
			<label class="label">Privacy</label>
			<p class="control select">
				<select v-model="newStation.privacy">
					<option value="public">Public</option>
					<option value="unlisted">Unlisted</option>
					<option value="private">Private</option>
				</select>
			</p>
		</template>
		<template #footer>
			<a class="button is-primary" @click="submitModal()">Create</a>
		</template>
	</modal>
</template>

<style lang="less" scoped>
.station-id {
	text-transform: lowercase;

	&::placeholder {
		text-transform: none;
	}
}
</style>
