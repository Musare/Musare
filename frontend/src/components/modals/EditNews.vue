<script setup lang="ts">
import { defineAsyncComponent, ref, onMounted } from "vue";
import { marked } from "marked";
import DOMPurify from "dompurify";
import Toast from "toasters";
import { formatDistance } from "date-fns";
import { useWebsocketStore } from "@/stores/websocket";
import { useModalsStore } from "@/stores/modals";
import { useForm } from "@/composables/useForm";

const Modal = defineAsyncComponent(() => import("@/components/Modal.vue"));
const SaveButton = defineAsyncComponent(
	() => import("@/components/SaveButton.vue")
);
const UserLink = defineAsyncComponent(
	() => import("@/components/UserLink.vue")
);

const props = defineProps({
	modalUuid: { type: String, required: true },
	createNews: { type: Boolean, default: false },
	newsId: { type: String, default: null },
	sector: { type: String, default: "admin" }
});

const { onReady, runJob } = useWebsocketStore();

const { closeCurrentModal } = useModalsStore();

const createdBy = ref();
const createdAt = ref(0);

const getTitle = () => {
	let title = "";
	const preview = document.getElementById("preview");

	// validate existence of h1 for the page title

	if (!preview || preview.childNodes.length === 0) return "";

	if (preview.childNodes[0].nodeName !== "H1") {
		for (let node = 0; node < preview.childNodes.length; node += 1) {
			if (preview.childNodes[node].nodeName) {
				if (preview.childNodes[node].nodeName === "H1")
					title = preview.childNodes[node].textContent || "";

				break;
			}
		}
	} else title = preview.childNodes[0].textContent || "";

	return title;
};

const { inputs, save, setOriginalValue } = useForm(
	{
		markdown: {
			value: "# Header\n## Sub-Header\n- **So**\n- _Many_\n- ~Points~\n\nOther things you want to say and [link](https://example.com).\n\n### Sub-Sub-Header\n> Oh look, a quote!\n\n`lil code`\n\n```\nbig code\n```\n",
			validate: (value: string) => {
				if (value === "") return "News item cannot be empty.";
				if (!getTitle())
					return "Please provide a title (heading level 1) at the top of the document.";
				return true;
			}
		},
		status: "published",
		showToNewUsers: false
	},
	({ status, messages, values }, resolve, reject) => {
		if (status === "success") {
			const query = {
				title: getTitle(),
				markdown: values.markdown,
				status: values.status,
				showToNewUsers: values.showToNewUsers
			};

			runJob(`data.news.${props.createNews ? "create" : "updateById"}`, {
				_id: props.createNews ? null : props.newsId,
				query
			})
				.then(resolve)
				.catch(reject);
		} else {
			if (status === "unchanged") new Toast(messages.unchanged);
			else if (status === "error")
				Object.values(messages).forEach(message => {
					new Toast({ content: message, timeout: 8000 });
				});
			resolve();
		}
	},
	{
		modalUuid: props.modalUuid
	}
);

onMounted(async () => {
	marked.use({
		renderer: {
			table(header, body) {
				return `<table class="table">
				<thead>${header}</thead>
				<tbody>${body}</tbody>
				</table>`;
			}
		}
	});

	await onReady(async () => {
		if (props.newsId && !props.createNews) {
			const data = await runJob(`data.news.findById`, {
				_id: props.newsId
			}).catch(() => {
				new Toast("News with that ID not found.");
				closeCurrentModal();
			});

			setOriginalValue({
				markdown: data.markdown,
				status: data.status,
				showToNewUsers: data.showToNewUsers
			});

			createdBy.value = data.createdBy;
			createdAt.value = data.createdAt;
		}
	});
});
</script>

<template>
	<modal
		class="edit-news-modal"
		:title="createNews ? 'Create News' : 'Edit News'"
		:size="'wide'"
		:split="true"
	>
		<template #body>
			<div class="left-section">
				<p><strong>Markdown</strong></p>
				<textarea v-model="inputs['markdown'].value"></textarea>
			</div>
			<div class="right-section">
				<p><strong>Preview</strong></p>
				<div
					class="news-item"
					id="preview"
					v-html="
						DOMPurify.sanitize(marked(inputs['markdown'].value))
					"
				></div>
			</div>
		</template>
		<template #footer>
			<div>
				<p class="control select">
					<select v-model="inputs['status'].value">
						<option value="draft">Draft</option>
						<option value="published" selected>Publish</option>
					</select>
				</p>

				<p class="is-expanded checkbox-control">
					<label class="switch">
						<input
							type="checkbox"
							id="show-to-new-users"
							v-model="inputs['showToNewUsers'].value"
						/>
						<span class="slider round"></span>
					</label>

					<label for="show-to-new-users">
						<p>Show to new users</p>
					</label>
				</p>

				<save-button
					ref="saveButton"
					v-if="createNews"
					@clicked="save()"
				/>

				<save-button
					ref="saveAndCloseButton"
					default-message="Save and close"
					@clicked="save(closeCurrentModal)"
				/>
				<div class="right" v-if="createdAt > 0">
					<span>
						By
						<user-link
							:user-id="createdBy"
							:alt="createdBy"
						/> </span
					>&nbsp;<span :title="new Date(createdAt).toString()">
						{{
							formatDistance(createdAt, new Date(), {
								addSuffix: true
							})
						}}
					</span>
				</div>
			</div>
		</template>
	</modal>
</template>

<style lang="less">
.edit-news-modal .modal-card .modal-card-foot .right {
	column-gap: 5px;
}
</style>

<style lang="less" scoped>
.night-mode {
	.edit-news-modal .modal-card .modal-card-body textarea,
	.edit-news-modal .modal-card .modal-card-body #preview {
		border-color: var(--grey-3);
	}

	.edit-news-modal .modal-card .modal-card-body textarea {
		background-color: var(--dark-grey);
		color: var(--white);
	}
}
.edit-news-modal .modal-card .modal-card-body {
	.left-section,
	.right-section {
		padding: 10px;
	}

	textarea {
		border: 0;
		outline: none;
		resize: none;
		font-size: 16px;
	}

	#preview {
		word-break: break-all;
		overflow: auto;
		box-shadow: 0;
	}

	textarea,
	#preview {
		padding: 5px;
		border: 1px solid var(--light-grey-3) !important;
		border-radius: @border-radius;
		height: calc(100vh - 280px);
		width: 100%;
	}
}

.edit-news-modal .modal-card .modal-card-foot {
	.control {
		margin-bottom: 0 !important;
	}

	.right {
		line-height: 36px;
		column-gap: 0;
	}
}
</style>
