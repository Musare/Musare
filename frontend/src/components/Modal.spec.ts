import { flushPromises } from "@vue/test-utils";
import { h } from "vue";
import { getWrapper } from "@/tests/utils/utils";
import { useConfigStore } from "@/stores/config";
import { useModalsStore } from "@/stores/modals";
import Modal from "@/components/Modal.vue";

describe("Modal component", () => {
	beforeEach(async context => {
		context.wrapper = await getWrapper(Modal);
	});

	test("title prop", async ({ wrapper }) => {
		const title = "Modal Title";
		await wrapper.setProps({ title });
		expect(wrapper.find(".modal-card-title").text()).toBe(title);
	});

	test("size prop", async ({ wrapper }) => {
		await wrapper.setProps({ size: "slim" });
		expect(wrapper.find(".modal-card").classes()).toContain("modal-slim");

		await wrapper.setProps({ size: "wide" });
		expect(wrapper.find(".modal-card").classes()).toContain("modal-wide");
	});

	test("split prop", async ({ wrapper }) => {
		expect(wrapper.find(".modal-card").classes()).not.toContain(
			"modal-split"
		);
		await wrapper.setProps({ split: true });
		expect(wrapper.find(".modal-card").classes()).toContain("modal-split");
	});

	test("christmas lights render if enabled", async () => {
		const wrapper = await getWrapper(Modal, {
			shallow: true
		});
		const configStore = useConfigStore();
		configStore.christmas = true;
		await flushPromises();
		expect(
			wrapper.findComponent({ name: "ChristmasLights" }).exists()
		).toBeTruthy();
	});

	test("click to close modal calls store action", async ({ wrapper }) => {
		const modalsStore = useModalsStore();
		await wrapper.find(".modal-background").trigger("click");
		await wrapper.find(".modal-card-head > .delete").trigger("click");
		await flushPromises();
		expect(modalsStore.closeCurrentModal).toHaveBeenCalledTimes(2);
	});

	test("renders slots", async () => {
		const wrapper = await getWrapper(Modal, {
			slots: {
				sidebar: h("div", {}, "Sidebar Slot"),
				toggleMobileSidebar: h("div", {}, "Toggle Mobile Sidebar Slot"),
				body: h("div", {}, "Body Slot"),
				footer: h("div", {}, "Footer Slot")
			}
		});
		expect(wrapper.html()).toMatchSnapshot();
	});
});
