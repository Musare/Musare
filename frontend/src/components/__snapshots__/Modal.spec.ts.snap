// Vitest Snapshot v1

exports[`Modal component > renders slots 1`] = `
"<div class=\\"modal is-active\\">
  <div class=\\"modal-background\\"></div>
  <div>Sidebar Slot</div>
  <div class=\\"modal-card\\">
    <header class=\\"modal-card-head\\">
      <div>Toggle Mobile Sidebar Slot</div>
      <h2 class=\\"modal-card-title is-marginless\\">Modal</h2><span class=\\"delete material-icons\\">highlight_off</span>
      <transition-stub appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
        <!--v-if-->
      </transition-stub>
    </header>
    <section class=\\"modal-card-body\\">
      <div>Body Slot</div>
    </section>
    <footer class=\\"modal-card-foot\\">
      <div>Footer Slot</div>
    </footer>
  </div>
</div>"
`;
